tests/framework/utils.py

```diff
@@ -450,16 +451,38 @@ def run_guest_cmd(ssh_connection, cmd, expected, use_json=False):
     assert stdout == expected
 
 
-@retry(wait=wait_fixed(1), stop=stop_after_attempt(10), reraise=True)
+def get_process_pidfd(pid):
+    """Get a pidfd file descriptor for the process with PID `pid`
+
+    Will return a pid file descriptor for the process with PID `pid` if it is
+    still alive. If the process has already exited it will return `None`.
+
+    Any other error while calling the system call, will raise an OSError
+    exception.
+    """
+    try:
+        pidfd = os.pidfd_open(pid)
+    except ProcessLookupError:
+        return None
+
+    return pidfd
+
+
 def wait_process_termination(p_pid):
     """Wait for a process to terminate.
 
-    Will return sucessfully if the process
+    Will return successfully if the process
     got indeed killed or raises an exception if the process
     is still alive after retrying several times.
     """
-    if psutil.pid_exists(p_pid):
-        raise Exception(f"[{p_pid}] process is still alive")
+    pidfd = get_process_pidfd(p_pid)
+
+    # If pidfd is None the process has already terminated
+    if pidfd is not None:
+        epoll = select.epoll()
+        epoll.register(pidfd, select.EPOLLIN)
```

Comment:
if the process goes between `get_process_pidfd` and `register`, will we still receive the `EPOLLIN`?

Reply:
I _think_ so. IIUC, if `pidfd_open` succeeds it opens a pidfd file and installs the file descriptor in the caller's file descriptor table. This means that we can poll on it until we close it.

Also, when we `poll` on it, we end up doing this: https://elixir.bootlin.com/linux/v6.11.3/source/fs/pidfs.c#L109, which checks for the `struct task` that corresponds to the `pid` that corresponds to the `pidfd`. If this is NULL, the system call returns `EPOLLIN` directly.

Reply:
I've changed this to use `select.select` directly